### PR TITLE
Add support for unstable feature advertising via /versions

### PR DESCRIPTION
**Pull request by turt2live**
Rendered: see 'docs' status check
Proposal: https://github.com/matrix-org/matrix-doc/issues/1497

----

The only potential change (it is unclear in the proposal) is that the spec doesn't define the features that can be listed here. The rationale being that people can be courteous to each other and not pick identifiers which collide with other people's implementation where possible.

### CODE REVIEW

**Comment by turt2live** [comment]
I've asked for the review of anyone on the core team, but would appreciate a review from @ara4n (to ensure accuracy with the proposal) and at least 1 other person.

File: api/client-server/versions.yaml [hunk 1]
```diff
@@ -33,13 +33,25 @@ paths:
...
         i.e. if the server implements ``r0.0.0``, ``r0.0.1``, and ``r1.2.0``, it will report ``r0.0.1`` and ``r1.2.0``.
+
+        The server may additionally advertise experimental features it supports
+        through ``unstable_features``. These features should be namespaced and
+        may optionally include version information within their name if desired.
+        Features listed here are not for optionally toggling parts of the Matrix
+        specification and should only be used to advertise support for a feature
+        which has not yet landed in the spec. For example, an accepted proposal
+        for a feature needing implementation would advertise itself here with
+        the intention of being removed from this list once the spec changes land.
```
Comment (KitsuneRal) [inline api/client-server/versions.yaml]: I feel uneasy about this part. Assume I have Quaternion 0.5 that chooses to support the (so far) unstable feature "org.matrix.kitchen_sink" if the server provides it. Quaternion 0.5 checks for this feature in `getVersions` output, things are fine so far. Now a new version of the spec gets released with "org.matrix.kitchen_sink" pulled in; according to the proposed text, the next version of Synapse drops "org.matrix.kitchen_sink" from the list of unstable features. Quaternion 0.5 keeps checking for the unstable feature though, and disables it for homeservers running the newer Synapse.

In other words, we have to define some forward-compatibility mechanism for clients here. One idea that comes to mind is turning items inside `versions` to objects and adding `unstable_features` as another key inside those objects.

Comment (jcgruenhage) [reply api/client-server/versions.yaml]: Clients could ask the server for a list of new supported features for a new spec version that it doesn't know about yet.

Comment (jcgruenhage) [reply api/client-server/versions.yaml]: (That would of course need a new endpoint)

Comment (turt2live) [reply api/client-server/versions.yaml]: This is how the proposal is worded, but I also don't mind softening the text to encourage servers to consider the clients that are using the features (ie: advertise the feature while also supporting it for some amount of time). Anything more complicated than that would certainly need a dedicated proposal.

@ara4n @KitsuneRal how does softening the language sound? (and does it feel acceptable to do that at this stage?)

Comment (KitsuneRal) [reply api/client-server/versions.yaml]: Upon some thinking - it's probably not so bad that it's coming that way: we don't want to use this as a full-fledged feature-testing mechanism after-all.

I'm fine if there's a recommendation to keep the record in `unstable_features` for some (undefined) time after the feature had landed so that client authors have time to update their code. At the same time (and that's probably the right thing, and I wouldn't mind speccing it) client authors can be advised to _not_ release stable versions of their applications that test servers for unstable features. Somewhat pushes us into speccing unstable features once they really are stable.

Comment (ara4n) [reply api/client-server/versions.yaml]: > I'm fine if there's a recommendation to keep the record in unstable_features for some (undefined) time after the feature had landed so that client authors have time to update their code. 

this sounds very sane to me and solves the problem nicely. i don’t see a problem modifying the wording to encompass this at this stage

Comment (turt2live) [reply api/client-server/versions.yaml]: I've changed the wording here to make it clearer that the server is responsible for taking the feature off the list on its own accord, and that it should make a reasonable decision when doing that.

@ara4n @KitsuneRal please take another look.

**Comment by turt2live** [comment]
@ara4n @KitsuneRal I'm not sure if I should take the silence as approving or busy - please let me know either way. Rich's approving review is appealing, but I also don't want people's opinions to be lost.

File: api/client-server/versions.yaml [hunk 2]
```diff
@@ -39,9 +39,12 @@ paths:
...
         specification and should only be used to advertise support for a feature
-        which has not yet landed in the spec. For example, an accepted proposal
-        for a feature needing implementation would advertise itself here with
-        the intention of being removed from this list once the spec changes land.
+        which has not yet landed in the spec. For example, a feature currently
+        undergoing the proposal process may appear here and eventually be taken
+        off this list once the feature lands in the spec and the server deems it
+        reasonable to do so. Servers may wish to keep advertising features here
+        after they've been released into the spec to give clients a chance to
+        upgrade appropriately.
```
Comment (KitsuneRal) [inline api/client-server/versions.yaml]: Can we also discourage clients from including `unstable_version` _checks_ to their stable releases?

Comment (erikjohnston) [reply api/client-server/versions.yaml]: (Though its probably fine if e.g. releases a stable version that uses unstable features behind beta lab flags or something)

Comment (turt2live) [reply api/client-server/versions.yaml]: How does "Additionally, clients should avoid using unstable features in their stable releases" sound?

**Comment by ara4n** [comment]
lgtm

File: api/client-server/versions.yaml (hunk 2 above)
Comment (KitsuneRal) [reply api/client-server/versions.yaml]: LGTM here too
